base-containers/base/inginious_container_api/run_student.py

```diff
@@ -221,6 +154,165 @@ def _hack_signals(receive_signal):
+        connection.sendmsg([b'S'], [(socket.SOL_SOCKET, socket.SCM_RIGHTS, array.array("i", [stdin, stdout, stderr]))])
+        connection.send(msgpack.dumps(
+            {"type": "run_student_command", "student_container_id": student_container_id, "command": cmd,
+             "teardown_script": teardown_script, "working_dir": working_dir, "ssh": ssh, "user": user}))
+        return connection
+    else:
+        # Send the command to the student_container via the agent
+        zmq_socket.send(msgpack.dumps(
+            {"type": "run_student_command", "socket_id": socket_id, "student_container_id": student_container_id, "command": cmd,
+             "teardown_script": teardown_script, "working_dir": working_dir,
+             "ssh": ssh, "user": user, "only_dockers": both_dockers}, use_bin_type=True))
+        zmq_socket.recv() #ignore answer
+        return None
+
+
+def allow_to_send_signals(signal_handler_callback, connection, student_container_id, both_dockers):
+    """ Allow to transfer signals """
+    if signal_handler_callback is not None:
+        if both_dockers:
+            def receive_signal(signum_s):  # send signal directly to student_container
+                signum_data = str(signum_s).zfill(3).encode("utf8")
+                connection.send(signum_data)
+        else:
+            def receive_signal(signum_s):  # send signal to student_container via docker agent
+                signum_data = str(signum_s).zfill(3).encode("utf8")
+                msg = {"type": "student_signal", "student_container_id": student_container_id, "signal_data": signum_data}
+                send_socket = zmq.asyncio.Context().socket(zmq.REQ)
+                send_socket.connect("ipc:///sockets/main.sock")
+                send_socket.send(msgpack.dumps(msg, use_bin_type=True))
+                send_socket.recv()
+        signal_handler_callback(receive_signal)
+
+
+def wait_until_finished(both_dockers, zmq_socket, stdin, stdout, stderr, student_container_id):
+    """ Dynamically handle stdin, stdout and stderr while waiting for final message """
+    # Start a process to handle the stdin and send it to the student_container
+    if not both_dockers:
+        stdin_handler = threading.Thread(target=handle_stdin, args=(stdin, student_container_id), daemon=True)
+        stdin_handler.start()
+
+    # handle the student_container outputs and wait for final message
+    message = None
+    msg_type = None
+    stdout_file = os.fdopen(stdout, 'w')
+    stderr_file = os.fdopen(stderr, 'w')
+
+    while msg_type != "run_student_retval":
+        zmq_socket.send(msgpack.dumps({"type": "run_student_ask_retval"}, use_bin_type=True))  # ping pong socket
+        message = msgpack.loads(zmq_socket.recv(), use_list=False, strict_map_key=False)
+        msg_type = message["type"]
+
+        if msg_type == "stdout":
+            stdout_file.write(message["message"].rstrip()+'\n')
+            stdout_file.flush()
+
+        if msg_type == "stderr":
```

Comment:
I believe it is better to handle both cases separately since both cases do not write in same file.
But don't hesitate to correct me if you have a cleaner code in mind !

Reply:
My bad, didn't read carefully.